spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BindableProxyFactory.java

```diff
@@ -127,30 +128,33 @@ public void doWith(Method method) throws IllegalArgumentException {
 				Output output = AnnotationUtils.findAnnotation(method, Output.class);
 				if (output != null) {
 					String name = BindingBeanDefinitionRegistryUtils.getChannelName(output, method);
-					validateChannelType(method.getReturnType());
-					MessageChannel sharedChannel = locateSharedChannel(name);
-					if (sharedChannel == null) {
-						BindableProxyFactory.this.outputHolders.put(name, new ChannelHolder(
-								BindableProxyFactory.this.channelFactory.createOutputChannel(name), true));
+					Class<?> returnType = method.getReturnType();
+					Object sharedBoundElement = locateSharedBoundElement(name, returnType);
+					if (sharedBoundElement != null) {
+						BindableProxyFactory.this.outputHolders.put(name,
+								new BoundElementHolder(sharedBoundElement, false));
 					}
 					else {
-						BindableProxyFactory.this.outputHolders.put(name, new ChannelHolder(sharedChannel, false));
+						BindableProxyFactory.this.outputHolders.put(name,
+								new BoundElementHolder(getBoundElementFactory(returnType).createOutput(name), true));
 					}
 				}
 			}
-
 		});
 	}
 
-	private void validateChannelType(Class<?> channelType) {
-		Assert.isTrue(SubscribableChannel.class.equals(channelType) || MessageChannel.class.equals(channelType),
-				"A bound channel should be either a '" + MessageChannel.class.getName() + "', " +
-						" or a '" + SubscribableChannel.class.getName() + "'");
+	private BindingTargetFactory getBoundElementFactory(Class<?> boundElementType) {
+		for (BindingTargetFactory factory : boundElementFactories) {
+			if (factory.canCreate(boundElementType)) {
```

Comment:
do we enforce only one factory per-type? (and if not, do we intend to support more than one per type?... this would always return the first one that returns `true`)